core/app/models/spree/order/address_book.rb

```diff
@@ -53,19 +53,20 @@ def ship_address_attributes=(attributes)
       def update_or_create_address(attributes = {})
         return if attributes.blank?
 
-        attributes.transform_values! { |v| v == '' ? nil : v }
+        attributes.transform_values!{ |v| v == "" ? nil : v }
 
-        address_scope = user ? user.addresses : ::Spree::Address
-        address = address_scope.find_by(id: attributes[:id])
+        address_scope = user ? user.addresses : ::Spree::Address.where(user: nil)
+        existing_address = address_scope.find_by(id: attributes[:id])
+        new_address = address_scope.new(attributes)
 
-        if address
-          address.update(attributes) if address.editable?
-
-          temp_address = address_scope.new(attributes)
-          return address if address == temp_address
+        if existing_address && new_address == existing_address
```

Comment:
This check and `new_address` is not needed. You can just do
```ruby
if existing_address
  existing_address. assign_attributes(attributes)
  existing_address.save if existing_address.changed?
else 
  address_scope.find_or_create_by(attributes.except(:id, :updated_at, :created_at))
end
```

Reply:
I think we still need to check, if the address is `.editable?`, otherwise someone could overwrite an existing guest order address which is already paid but not shipped yet.

Reply:
`new_address` actually also helps with another problem: if a user resubmits the exact same address, but said address is not `editable?`, then a copy will be created, which is not needed.

Reply:
What I meant is, you don't need to initialize a new object (`new_address`) and check for equality like that, you can just `assign_attributes` and check if `changed?`. If it's editable then just save, if it's not editable then create a new one with find_or_create_by.